ert_shared/ensemble_evaluator/evaluator.py

```diff
@@ -93,30 +94,39 @@ def create_snapshot_msg(ee_id, snapshot, event_index):
 
     async def handle_client(self, websocket, path):
         logger.debug(f"Client {websocket.remote_address} connected.")
-        self._clients.add(websocket)
-
-        message = self.create_snapshot_msg(
-            self._ee_id, self._snapshot, self.event_index()
-        )
-        await websocket.send(message)
 
-        async for message in websocket:
-            client_event = from_json(message)
-            if client_event["type"] == identifiers.EVTYPE_EE_TERMINATE_REQUEST:
-                logger.debug(f"Client {websocket.remote_address} asked to terminate.")
-                self._stop()
-        logger.debug(f"Client {websocket.remote_address} disconnected.")
-        self._clients.remove(websocket)
+        self._clients.add(websocket)
+        try:
+            message = self.create_snapshot_msg(
+                self._ee_id, self._snapshot, self.event_index()
+            )
+            await websocket.send(message)
+
+            async for message in websocket:
+                client_event = from_json(message)
+                if client_event["type"] == identifiers.EVTYPE_EE_TERMINATE_REQUEST:
+                    logger.debug(
+                        f"Client {websocket.remote_address} asked to terminate."
+                    )
+                    self._stop()
+            logger.debug(f"Client {websocket.remote_address} disconnected.")
+        finally:
+            self._clients.remove(websocket)
 
     async def handle_dispatch(self, websocket, path):
         logger.debug(f"Dispatch {websocket.remote_address} connected.")
-        async for msg in websocket:
-            logger.debug(f"Dispatch got: {msg}.")
-            if msg == "null":
-                return
-            event = from_json(msg)
-            await dispatch.handle_event(self, event)
-        logger.debug(f"Dispatch {websocket.remote_address} disconnected.")
+        await self._dispatchers_connected.put(None)
+        try:
+            async for msg in websocket:
+                logger.debug(f"Dispatch got: {msg}.")
+                if msg == "null":
+                    return
+                event = from_json(msg)
+                await dispatch.handle_event(self, event)
+            logger.debug(f"Dispatch {websocket.remote_address} disconnected.")
+        finally:
+            await self._dispatchers_connected.get()
+            self._dispatchers_connected.task_done()
```

Comment:
Same here. `try` with only a `finally` will probably cause confusion? At least I am confused, because I have to google in order to figure out what happens to the exception.

Reply:
Replaced both with a context manager. Hope it makes the code clearer.